srk/src/chc.ml

```diff
@@ -0,0 +1,408 @@
+    Format.fprintf formatter "@]\nQueries:@[%a@])"
+    (SrkUtil.pp_print_enum_nobox (pp_query rel_ctx)) 
+      (BatList.enum chc.queries)
+
+
+  let show srk rel_ctx = SrkUtil.mk_show (pp srk rel_ctx)
+
+  let create = {rules=[];queries=[]}
+  let chc_of rules queries = {rules; queries}
+
+  let add_rule chc rule = {rules=rule::chc.rules; queries=chc.queries} 
+  let add_query chc query = {rules=chc.rules; queries=query::chc.queries}
+  let get_rules chc = chc.rules
+  let get_queries chc = chc.queries
+
+  let get_relations_used chc =
+    List.fold_left
+      (fun rset ((hypo_atoms, _), conc_atom) ->
+         List.fold_left
+           (fun rset rel_atom -> Relation.Set.add (rel_of_atom rel_atom) rset)
+           (Relation.Set.add (rel_of_atom conc_atom) rset)
+           hypo_atoms)
+      (List.fold_left 
+         (fun rset query -> Relation.Set.add query rset)
+         Relation.Set.empty
+         chc.queries)
+      chc.rules
+
+  let is_linear chc =
+    BatList.fold_left 
+      (fun acc ((rel_atoms, _), _) -> acc && (List.length rel_atoms) <= 1)
+      true
+      chc.rules
+
+  let goal_vert = -2 
+  let start_vert = -1 
+  let to_weighted_graph srk chc pd =
+    let open WeightedGraph in
+    let emptyarr = BatArray.init 0 (fun _ -> failwith "empty") in
+    let alg = 
+      let map_union m1 m2 = Symbol.Map.fold Symbol.Map.add m1 m2 in
+      let mk_subst_map a b =
+        BatArray.fold_lefti 
+          (fun map ind asym -> Symbol.Map.add asym (mk_const srk (b.(ind))) map)
+          Symbol.Map.empty
+          a
+      in
+      let is_one (_, _, phi) = Formula.equal phi (mk_true srk) in
+      let is_zero (_, _, phi) = Formula.equal phi (mk_false srk) in
+      let zero = emptyarr, emptyarr, mk_false srk in
+      let one = emptyarr, emptyarr, mk_true srk in
+      let add x y =
+        if is_zero x then y else if is_zero y then x
+        else if is_one x || is_one y then one
+        else (
+          let (pre, post, _) = x in
+          let pre' = Array.map (fun sym -> dup_symbol srk sym) pre in
+          let post' = Array.map (fun sym -> dup_symbol srk sym) post in
+          let subst (pre, post, phi) = 
+            substitute_map
+              srk 
+              (map_union (mk_subst_map pre pre') (mk_subst_map post post'))
+              phi
+          in
+          pre', post', mk_or srk [subst x; subst y])
+      in
+      let mul x y = 
+        if is_zero x || is_zero y then zero
+        else if is_one x then y else if is_one y then x
+        else (
+          let (pre1, post1, phi1) = x in
+          let (pre2, post2, phi2) = y in
+          let pre' = Array.map (fun sym -> dup_symbol srk sym) pre1 in
+          let post' = Array.map (fun sym -> dup_symbol srk sym) post2 in 
+          let rhs_pre_postmap = 
+            map_union (mk_subst_map post2 post') (mk_subst_map pre2 post1)
+          in
+          let rhs_subst =
+            Memo.memo 
+              ~size:(Symbol.Set.cardinal (symbols phi2))
```

Comment:
Is this true even when I know the exact number of buckets ahead of time (as I do in this case)? It seems like when you know exact number there is no reason to over-estimate (assuming the table does not change size until it needs to).

Reply:
You always want to over-estimate, unless you have a perfect hash function.  Even if you do have a perfect hash function, you may still want to over-estimate if you're using a hash table from a library, since that hash table will almost certainly resize the underlying table when it gets close to being full.